File: repos/sample-ts/src/auth/AuthService.ts
import { Validator } from "./Validator";
import { checkCreditLimit, explainCreditDecision } from "../credit/limits";
import { Customer } from "../credit/Customer";

/**
 * Service xử lý đăng nhập + xét duyệt khoản vay nhanh.
 * Login đơn giản chỉ check format; quyết định cho vay dựa trên hạn mức tín dụng.
 */
export class AuthService {
  constructor(private validator: Validator) {}

  login(username: string, password: string): boolean {
    if (!this.validator.validate(username, password)) {
      return false;
    }
    return this.lookupActiveSession(username) !== null;
  }

  /** Khách đã đăng nhập có thể yêu cầu khoản vay; trả lý do nếu bị từ chối. */
  approveLoan(customer: Customer, requestedAmount: number): { ok: boolean; reason: string } {
    const ok = checkCreditLimit(customer, requestedAmount);
    return { ok, reason: ok ? "Approved" : explainCreditDecision(customer, requestedAmount) };
  }

  private lookupActiveSession(username: string): string | null {
    return username ? `session-${username}` : null;
  }
}

File: repos/sample-cs/src/Auth/AuthService.cs
using SampleApp.Credit;

namespace SampleApp.Auth;

/// <summary>Service đăng nhập + xét duyệt vay; entry point của module Auth.</summary>
public class AuthService : IAuthService
{
    private readonly Validator _validator;
    private readonly CreditChecker _creditChecker;
    private readonly Dictionary<string, Customer> _customers;

    public AuthService(Validator validator, CreditChecker creditChecker)
    {
        _validator = validator;
        _creditChecker = creditChecker;
        _customers = new Dictionary<string, Customer>();
    }

    public bool Login(string username, string password)
    {
        return _validator.Validate(username, password);
    }

    /// <summary>Khách đã đăng nhập có thể yêu cầu khoản vay.</summary>
    public bool ApproveLoan(string customerId, decimal requestedAmount)
    {
        if (!_customers.TryGetValue(customerId, out var customer))
        {
            return false;
        }
        return _creditChecker.CheckLimit(customer, requestedAmount);
    }

    public void RegisterCustomer(Customer customer)
    {
        _customers[customer.Id] = customer;
    }
}
